pkg/apis/pipeline/v1alpha1/task_validation.go

```diff
@@ -132,67 +132,110 @@ func validateInputParameterVariables(steps []corev1.Container, inputs *Inputs) *
 			parameterNames[p.Name] = struct{}{}
 		}
 	}
-	return validateVariables(steps, "params", parameterNames)
+	return validateVariables(steps, "params", "inputs.", parameterNames)
 }
 
 func validateResourceVariables(steps []corev1.Container, inputs *Inputs, outputs *Outputs) *apis.FieldError {
-	resourceNames := map[string]struct{}{}
+	inputVars, err := getInputResourceVariables(inputs)
+	if err != nil {
+		return err
+	}
+	outputVars, err := getOutputResourceVariables(outputs)
+	if err != nil {
+		return err
+	}
+	err = validateVariables(steps, "resources", "inputs.", inputVars)
+	if err != nil {
+		return err
+	}
+	return validateVariables(steps, "resources", "outputs.", outputVars)
+}
+
+func getInputResourceVariables(inputs *Inputs) (map[string]struct{}, *apis.FieldError) {
```

Comment:
I'm curious why you chose to use two separate functions for input & output here - it seems like it would be possible to write a single function to handle both with a signature like:

```go
func getResourceVariables(resources []TaskResource, pathPrefix string)
```

I might be misreading but it looks like the only line of difference between `getInputResourceVariables` and `getOutputResourceVariables` is the path string in the returned `apis.FieldError`, which could be resolved using a `pathPrefix` arg.

Reply:
You're totally right! There was plenty of duplicate code there. I went ahead and made these changes. Thanks for the suggestion.